typing/typemod.ml

```diff
@@ -1847,16 +1847,28 @@ let rec nongen_modtype env = function
 and nongen_signature_item env = function
     Sig_value(_id, desc, _) -> Ctype.nongen_schema env desc.val_type
   | Sig_module(_id, _, md, _, _) -> nongen_modtype env md.md_type
-  | _ -> false
+  | _ -> None
 
 let check_nongen_signature_item env sig_item =
   match sig_item with
     Sig_value(_id, vd, _) ->
-      if Ctype.nongen_schema env vd.val_type then
-        raise (Error (vd.val_loc, env, Non_generalizable vd.val_type))
+      begin match Ctype.nongen_schema env vd.val_type with
+      | None -> ()
+      | Some variable ->
+          let error =
+            Non_generalizable { variable; expression = vd.val_type }
+          in
+          raise (Error (vd.val_loc, env, error))
+      end
   | Sig_module (_id, _, md, _, _) ->
-      if nongen_modtype env md.md_type then
-        raise(Error(md.md_loc, env, Non_generalizable_module md.md_type))
+      begin match nongen_modtype env md.md_type with
+      | Some variable ->
+          let error =
+            Non_generalizable_module { variable; mty = md.md_type }
```

Comment:
I have the impression that the most impactful change in term of usability improvement would be to return the precise signature item that made the module non generalizable by making `nongen_modtype` returns the first non-generalisable item. In this case, I think we could return only the first item because there could be a dependency relation between the first items and the remaining one. Thus it could useful to nudge users towards fixing this first non-generalizable item.